app/src/main/java/org/thoughtcrime/securesms/database/StickerDatabase.java

```diff
@@ -0,0 +1,532 @@
+
+    setNotifyStickerPackListeners(cursor);
+    return cursor;
+  }
+
+  public @Nullable Cursor getStickersByEmoji(@NonNull String emoji) {
+    String   selection = EMOJI + " LIKE ? AND " + COVER + " = ?";
+    String[] args      = new String[] { "%"+emoji+"%", "0" };
+
+    Cursor cursor = databaseHelper.getSignalReadableDatabase().query(TABLE_NAME, null, selection, args, null, null, null);
+    setNotifyStickerListeners(cursor);
+
+    return cursor;
+  }
+
+  public @Nullable Cursor getAllStickerPacks() {
+    return getAllStickerPacks(null);
+  }
+
+  public @Nullable Cursor getAllStickerPacks(@Nullable String limit) {
+    String   query  = COVER + " = ?";
+    String[] args   = new String[] { "1" };
+    Cursor   cursor = databaseHelper.getSignalReadableDatabase().query(TABLE_NAME, null, query, args, null, null, PACK_ORDER + " ASC", limit);
+    setNotifyStickerPackListeners(cursor);
+
+    return cursor;
+  }
+
+  public @Nullable Cursor getStickersForPack(@NonNull String packId) {
+    SQLiteDatabase db        = databaseHelper.getSignalReadableDatabase();
+    String         selection = PACK_ID + " = ? AND " + COVER + " = ?";
+    String[]       args      = new String[] { packId, "0" };
+
+    Cursor cursor = db.query(TABLE_NAME, null, selection, args, null, null, null);
+    setNotifyStickerListeners(cursor);
+
+    return cursor;
+  }
+
+  public @Nullable Cursor getRecentlyUsedStickers(int limit) {
+    SQLiteDatabase db        = databaseHelper.getSignalReadableDatabase();
+    String         selection = LAST_USED + " > ? AND " + COVER + " = ?";
+    String[]       args      = new String[] { "0", "0" };
+
+    Cursor cursor = db.query(TABLE_NAME, null, selection, args, null, null, LAST_USED + " DESC", String.valueOf(limit));
+    setNotifyStickerListeners(cursor);
+
+    return cursor;
+  }
+
+  public @NonNull Set<String> getAllStickerFiles() {
+    SQLiteDatabase db        = databaseHelper.getSignalReadableDatabase();
+
+    Set<String> files = new HashSet<>();
+    try (Cursor cursor = db.query(TABLE_NAME, new String[] { FILE_PATH }, null, null, null, null, null)) {
+      while (cursor != null && cursor.moveToNext()) {
+        files.add(CursorUtil.requireString(cursor, FILE_PATH));
+      }
+    }
+
+    return files;
+  }
+
+  public @Nullable InputStream getStickerStream(long rowId) throws IOException {
+    String   selection = _ID + " = ?";
+    String[] args      = new String[] { String.valueOf(rowId) };
+
+    try (Cursor cursor = databaseHelper.getSignalReadableDatabase().query(TABLE_NAME, null, selection, args, null, null, null)) {
+      if (cursor != null && cursor.moveToNext()) {
+        String path   = cursor.getString(cursor.getColumnIndexOrThrow(FILE_PATH));
+        byte[] random = cursor.getBlob(cursor.getColumnIndexOrThrow(FILE_RANDOM));
+
+        if (path != null) {
+          return ModernDecryptingPartInputStream.createFor(attachmentSecret, random, new File(path), 0);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code contains a resource that might not have closed properly. A resource leak can slow down or crash your system.

**Fix**
Consider closing the resource returned by the *createFor* method in a try-finally block. The resource is returned at line 222. There are other execution paths that don&#39;t close the resource or return it.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).